### PR TITLE
#165412918  Give a rating to an article

### PR DESCRIPTION
**Pull request by djallas**
#### What does this PR do?
Give a rating on the article written by others

#### Description of Task to be completed?
Have the following endpoint accessible to an authenticated user:

`POST /api/v1/articles/:slug/rating` give rating to an article identified by its slug
`GET /api/v1/articles/:slug/rating` Get users who rated a given article

#### How should this be manually tested?
- After cloning the **repo**, `cd` into it and type `npm test`
- Using Postman, 
->make sure you signup then login to get the access token, 
->create an article,
-> Create a rating using this endoint: `/api/v1/articles/:slug/rating`
-> Make sure you pass the access token in the header with the following key name: `access-token`,
-> Finally, send a **JSON** object of the following structure : {"rating":  1}

#### Any background context you want to provide?
- The formula used to calculate a rating of a given article
Given,
`n`:  number of rating per article
`s`: sum of ratings per a given article

```
rating: s/n
```

#### What are the relevant pivotal tracker stories?
[#165412918](https://www.pivotaltracker.com/story/show/165412918)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     "chance": "^1.0.18",
     "cheke": "^1.0.3",
     "cors": "^2.8.5",
+    "d": "^1.0.0",
```
Comment (e-liyai) [inline package.json]: what's this meant for?

File: src/middlewares/validation/validateRating.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  /**
+   * @param {object} req Request sent to the route
+   * @param {object} res Response from server
+   * @param {object} next If no error continue
+   * @returns {object} Object representing the response returned
+   */
+  static create(req, res, next) {
+    const errors = {};
+    if (req.body && typeof req.body.rating !== 'number') errors.rating = 'rating must be a number';
+    if (req.body.rating < 1) errors.rating = 'rating must be larger than or equal to 1';
```
Comment (e-liyai) [inline src/middlewares/validation/validateRating.js]: line 16 and 17 can be done by one check that looks at the range

File: src/queries/ratings/create.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * @param {object} rating input user rating
+ * @returns {object} Object representing the response returned
+ */
+
+export default async (data = {}) => {
+  const { articleId, userId, rating } = data;
+  // check if already exist
+  const result = await db.Rating.findAll({
```
Comment (e-liyai) [inline src/queries/ratings/create.js]: there should be a file for `findAll` Ratings then import it here

File: package.json (hunk 1 above)
Comment (djallas) [reply package.json]: The package is no longer needed, I am removing it

File: src/queries/ratings/update.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import db from '../../models';
+
+export default async (articleId) => {
+  const allRatings = await db.Rating.findAll({
```
Comment (e-liyai) [inline src/queries/ratings/update.js]: import this as well

File: src/models/rating.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import * as queries from '../queries';
```
Comment (e-liyai) [inline src/models/rating.js]: there should not be queries in a model file. Set static data

**Comment by djallas** [comment]
@e-liyai made changes, you can check.

**Comment by djallas** [comment]
@e-liyai check this PR is ready.

**Comment by djallas** [comment]
@e-liyai can you check this PR for me. it's ready 👍

File: src/queries/ratings/create.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+export default async (data = {}) => {
+  const { articleId, userId, rating } = data;
+  // check if already exist
+  const result = await findAll(articleId, userId);
+  const response = {};
+  if (result && result.length === 0) {
+    await db.Rating.create({ rating, articleId, userId }, { logging: false });
+    response.message = 'Thank you for rating this article';
```
Comment (e-liyai) [inline src/queries/ratings/create.js]: remove message in queries, it should be in controller

File: src/queries/ratings/create.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    response.statusCode = status.CREATED;
+  } else {
+    await db.Rating.update(
+      { rating },
+      { where: { [db.Op.and]: { articleId, userId } }, logging: false }
+    );
+    response.statusCode = status.OK;
+    response.message = 'Your article rating has been updated';
```
Comment (e-liyai) [inline src/queries/ratings/create.js]: same here

File: src/controllers/RatingController.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      rating: req.body.rating,
+      articleId: req.article.id,
+      userId: req.user.id
+    };
+    const response = await Article.rate.create(data);
+    return res.status(response.statusCode).send({
+      rating: {
+        message:
+          response.statusCode === status.OK
+            ? 'Your article rating has been updated'
```
Comment (e-liyai) [inline src/controllers/RatingController.js]: the logic here seems flawed if the status is ok `article update successful` else even in the case of an error the message displayed is `Thank you for rating this article`

File: src/queries/ratings/create.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * @param {object} rating input user rating
+ * @returns {object} Object representing the response returned
+ */
+
+export default async (data = {}) => {
+  const { articleId, userId, rating } = data;
+  // check if already exist
+  const result = await findAll(articleId, userId);
+  const response = {};
+  if (result && result.length === 0) {
```
Comment (e-liyai) [inline src/queries/ratings/create.js]: i see a create or update scenario here, update the filename appropriately

**Pull request merged**